### PR TITLE
Fix #1736: Add a scalajsp input task in the sbt plugin.

### LINKED ISSUE #1736: A scalajsp command/task in the sbt plugin
I find myself wanting this all the time when I work on the compiler/optimizer. It's only the only feature that the CLI provides and the sbt plugin doesn't.

### PR DESCRIPTION
**Pull request by sjrd**
It accepts relative paths to .sjsir files on the classpath.

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @gzm0

File: tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala [hunk 1]
```diff
@@ -61,9 +61,10 @@ trait JarTraverser extends ClasspathContentHandler with FileSystem {
     def entryBinaryContent: Array[Byte] = IO.readInputStreamToByteArray(in)
     def entryVersion: Option[String] = Some(entry.getTime.toString)
 
-    if (longName == JSDependencyManifest.ManifestFileName)
-      handleDepManifest(JSDependencyManifest.read(entryReader))
-    else if (longName.endsWith(".js")) {
+    if (longName == JSDependencyManifest.ManifestFileName) {
+      lazy val vf = JSDependencyManifest.read(entryReader)
+      handleDepManifest(vf)
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala]: Why this change? Do you just think the line is too long? I think in any case you should make this a def. We don't need the lazy behavior.

File: tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala [hunk 2]
```diff
@@ -74,7 +75,7 @@ trait JarTraverser extends ClasspathContentHandler with FileSystem {
       getOrCreateJSFile(relPath, fullPath, shortName)
         .withSourceMap(Some(entryContent))
     } else if (longName.endsWith(".sjsir")) {
-      val vf = new MemVirtualSerializedScalaJSIRFile(fullPath) {
+      lazy val vf = new MemVirtualSerializedScalaJSIRFile(fullPath) {
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala]: Same here. Make this a `def`

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala [hunk 3]
```diff
@@ -133,8 +142,77 @@ object ScalaJSPluginInternal {
...
+  private def filterOutReflProxies(tree: ir.Trees.ClassDef): ir.Trees.ClassDef = {
+    import ir.Trees._
+    import ir.Definitions.isReflProxyName
+    val newDefs = tree.defs.filter {
+      case MethodDef(_, Ident(name, _), _, _, _) => !isReflProxyName(name)
+      case _ => true
+    }
+    tree.copy(defs = newDefs)(tree.optimizerHints)(tree.pos)
+  }
+
```
Comment (gzm0) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala]: Have you considered moving this to `tools` or even `ir`? It seems it could be generally useful...

File: tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala (hunk 1 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala]: It's precisely to give lazy behavior, so that several accesses to the virtual file don't reload it every time.

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala (hunk 3 above)
Comment (sjrd) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala]: I did consider it, but I don't think it's generally useful. In theory one should never remove those. It's only useful for display. Besides, if they're in tools or ir, it becomes a public API, and I don't think this should be a public API, for the reasons outlined in the first sentence.

File: tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala (hunk 1 above)
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala]: I see... but shouldn't the callee handle that? It seems strange to assume that a callee unnecessarily evaluates a by name parameter multiple times...

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala (hunk 3 above)
Comment (gzm0) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala]: Ok.

File: tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala (hunk 1 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala]: But it's true that the lazy behavior is not really needed. And a client who needs it can build it on its side.

Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/classpath/builder/JarTraverser.scala]: So OK, I'll revert this line, and make the other one a def.

**Comment by sjrd** [comment]
Updated.

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala [hunk 4]
```diff
@@ -185,6 +185,11 @@ object ScalaJSPlugin extends AutoPlugin {
         "Private class loader to load jetty8 without polluting classpath. Only use this " +
         "as the `jettyClassLoader` argument of the PhantomJSEnv",
         KeyRanks.Invisible)
+
+    /** Prints the content of a .sjsir file in human readable form. */
+    val scalajsp = InputKey[Unit]("scalajsp",
+        "Prints the content of a .sjsir file in human readable form.",
+        DSetting)
```
Comment (gzm0) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala]: This should probably be a `DTask`.

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala [hunk 5]
```diff
@@ -133,8 +142,77 @@ object ScalaJSPluginInternal {
...
+        infos: Boolean = false,
+        showReflProxy: Boolean = false
+    )
+
+    val optionsParser: Parser[Options] = {
+      token(OptSpace ~> (
+          (literal("-i") | "--infos") ^^^ ((_: Options).copy(infos = true))
+        | (literal("-p") | "--reflProxies") ^^^ ((_: Options).copy(showReflProxy = true))
+      )).* map {
+        fns => fns.foldLeft(Options())((prev, fn) => fn(prev))
```
Comment (gzm0) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala]: Consider:

```scala
map { fns => Function.chain(fns)(Options()) }
```

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala [hunk 6]
```diff
@@ -3,8 +3,10 @@ package org.scalajs.sbtplugin
 import sbt._
 import sbt.inc.{IncOptions, ClassfileManager}
 import Keys._
-import sbinary.DefaultProtocol.StringFormat
+import sbinary.DefaultProtocol._
 import Cache.seqFormat
+import complete.Parser
+import complete.DefaultParsers
```
Comment (gzm0) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala]: This import is unused.

**Comment by gzm0** [comment]
That's all. Very nice!

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala (hunk 5 above)
Comment (sjrd) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala]: Oh neat. I didn't know that one!

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala (hunk 4 above)
Comment (sjrd) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala]: Yep. Made it a CTask actually (sbt says C is "advanced tasks", which I think fits).

**Comment by sjrd** [comment]
Updated.

**Comment by gzm0** [comment]
LGTM

**Comment by sjrd** [comment]
Updated with just a Scalastyle comment to let me use `return`.
